py-rse/package-py.Rmd

```diff
@@ -20,445 +20,110 @@ This lesson shows you how to use Python's tools to create and share libraries of
-printed?
-
-### Exercise: Turn `use.py` into a unit test
-
-1. Turn the `use.py` script into a `pytest` unit test; name the file `test_zipf.py`.
-
-## How do I create a package? {#py-rse-package-py-creating}
-
-Let's say we wanted to add some capabilities to what
-we've built: we want to be able to generate noisy
-Zipf distributions—noise within the same default `RELATIVE_ERROR`—and
-we want to add a Zipf-checker that can handle unsorted inputs:
-
-```python
-...as before...
-
-def make_zipf(length):
-    assert length > 0, 'Zipf distribution must have at least one element'
-    result = [1/(1 + i) for i in range(length)]
-    return result
-
-def make_noisy_zipf(length, rel=RELATIVE_ERROR):
-    data = make_zipf(length)
-    minnoise = 1.0 - rel/2
-    maxnoise = 1.0 + rel/2
-    for i in range(length):
-        data[i] = data[i] * random_uniform(minnoise, maxnoise)
-    return data
-
-def is_zipf(hist, rel=RELATIVE_ERROR):
-    assert len(hist) > 0, 'Cannot test Zipfiness without data'
-    scaled = [h/hist[0] for h in hist]
-    perfect = make_zipf(len(hist))
-    return scaled == approx(perfect, rel=rel)
-
-def is_unsorted_zipf(hist, rel=RELATIVE_ERROR):
-    sortedhist = sorted(hist)
-    return is_zipf(unsorted, rel)
-
-...as before...
-```
-
-As our package grows,
-we should split its source code into multiple files,
-and this is now starting to look like two broad types of
-related functionality: generating a Zipf distribution, and
-checking against a Zipf distribution.  But they are related;
-and what's more, there are constants that would be common
-across the two pieces - the `RELATIVE_ERROR` constant.
-
-If we want multiple files, but they are related and
-and have some common items, that's a good indication that
-we are outgrowing a single-file module and we want a Python
-package.
-
-A package in Python is no more than a directory containing
-Python code, and by convention a Python file with a special
-name, `__init__.py` (this was a requirement before Python 3.3).
+```text
+zipfs_law
+├── zipfs_law
+│   ├── __init__.py
+│   ├── collate.py
+│   ├── countwords.py
+│   ├── mymodule.py
+│   └── plotcounts.py
+├── README.md
+└── setup.py
+```
+
+It might initially be confusing
+that the top level directory has the same name as the package directory containing the `.py`-files.
+You could change the name of the top level folder to whatever you want,
+but it is often convenient that this is the same as the package directory.
+If you change the name of the package directory,
+this will change the name of your package.
+You could also place your package_directory inside a subdirectory,
+such as `src` or `py`,
+but this would require some changes to the setup instructions later on,
+so we will stick to the conventional folder hierarchy above.
+
+The folder contains a Python file with a special name: `__init__.py`.
+This was a requirement before Python 3.3,
+and is now included as a convention
+or to execute code upon package import.
```

Comment:
I think the original intent with this sentence is that some people just add an empty `__init__.py` because that is how it used to be and some people add an `__init__.py` specifically to execute the code it contains on import.